rfcs/0038-unified-device-target-and-memory-scope-planning.md

```diff
@@ -0,0 +1,244 @@
+   `device_copy` operation is required between those scopes. We'll pick this issue up again after
+   [RFC #9](https://github.com/apache/tvm-rfcs/blob/main/rfcs/0009_Unified_Static_Memory_Planning.md)
+   has landed.
+
+3. The `on_device` and `device_copy` call attributes use `SEScope`s instead of integers. However the Python
+   bindings for these 'operators' continue to accept a `Device` for convenience. The machinery in `LowerTEPass`
+   which resolves `DLDeviceTypes` to `Targets` is moved up in the compilation flow and becomes part of
+   `PlanDevices`. In particular, any `SEScope` encountered during device planning is 'canonicalized' to fill
+   in a `Target` by the same lookup as we do today. This means we continue to support the easy shorthand of
+   referring to devices by the `DLDeviceType` alone. However, advanced users can supply a `SEScope` to these
+   operators which contains the exact `Target` to use.
+
+4. We rework device planning to be in terms of `SEScope`s instead of `DLDeviceTypes`. Two `SEScope`s
+   become special:
+    - We need a default scope for all primitive operators which are not otherwise
+      constrained to a particular scope.
+    - We need a scope for 'host-only' operations and data, such as for shapes and shape functions.
+      (Currently this is hardcoded to `kDLCPU`).
+
+5. We extend `PlanDevices` to be able to a) run *after* lowering and b) refine existing constraints.  It will
+   look inside calls to `PrimFunc`s and follow the chain:
+
+   ```
+   tir::PrimFunc.buffer_map -> tir::Buffer.data -> tir::Var.type_annotation -> PointerType.storage_scope -> String
+   ```
+
+   to discover the memory scope for each Relay argument. That scope will enter `SEScope`s and flow through the
+   existing unification machinery. The existing sub-pass in `PlanDevices` will insert `device_copy` calls
+   wherever sub-expressions disagree on their memory scope.
+
+   (An additional pass is planned to heuristically move `device_copy`s around, and eliminate redundant
+    copies, however that's outside the scope of this RFC.)
+
+6. We rework `PartitionGraph` to `PartitionBySEScope` to work on `SEScope` annotations instead of
+   `compiler_begin` and `compiler_end` annotations. Algorithmically it's not a big change -- maximal
+   sub-expressions which share the same `SEScope` (or a projection thereof, eg just the `target`) are hoisted
+   into global `Function`s. The function's `"result_se_scope"` attribute describes both the scope holding the
+   function's result *and* the `Target` for which the function is to be compiled.
+
+7. We allow `MergeComposite` to be used to insert `on_device` annotations, call it `MergeAndAnnotate`.
+
+8. (?) We rework `AnnotateTarget` to just look for `FTVMAnnotateTarget` operator attributes, call it
+   `AnnotateSEScopes`. When the function fires an `on_device` annotation is inserted. However since
```

Comment:
clarifying my understanding:
```suggestion
   `AnnotateSEScopes`. When `FTVMAnnotateSEScopes` returns true, an `on_device` annotation is inserted. However since
```